DESIGN.md

```diff
@@ -0,0 +1,364 @@
+Pirouette's Design
+==================
+
+This document aims at giving an overview of the crucial parts of Pirouette, and
+pointers to get familiar with the code. As such, it contains three parts.
+
+1. The [first part][languages] describe how Pirouette supports several input
+   languages as well as two languages that are particularly relevant to us: [the
+   example language] and the support for Plutus IR – [Plutus intermediate
+   representation].
+
+2. The [second part][transformations] describes the transformations that
+   Pirouette performs on the input language to get it in a shape ready for
+   symbolic execution.
+
+3. The [third part][symbolic engine] digs into what constitutes the heart of
+   Pirouette: its symbolic engine, and in particular the [symbolic evaluator]
+   and the [symbolic prover].
+
+Languages
+---------
+
+[languages]: #languages
+
+### Pirouette's genericity
+
+Pirouette attempts to be generic in the input language by only assuming that it
+is System F-like. Anyone can then come and define their instance of
+`LanguageBuiltins`, which is Pirouette's way of defining the constants, builtin
```

Comment:
I think it would deserve further documentation because it did not strike me as easy to figure out, no. However, I haven't made the effort to figure it out and I am not sure what I could add to the document with my current knowledge. Maybe we simply open an issue stating that we want more details about this?

Reply:
Opening an issue works, and maybe leaving a TODO note.

Reply:
cf https://github.com/tweag/pirouette/issues/189.